Implement real-time PP counter

smoogipoo opened the pull request:
Preview: https://drive.google.com/file/d/1vOUF6uXfSNRJf2Onb7tsLjVUwawpoHny/view?usp=sharing

May not work super well with osu!lazer scores (without classic mod), especially visible in mania which displays 0pp until 500k score. The difficulty calculators are all currently tuned for osu!stable.

This creates a new `PerformanceCalculator` with every new judgement. From my testing, I've found the cost of doing this to be quite small, on the order of a couple hundred KB over the course of a map:
![osu](https://user-images.githubusercontent.com/1329837/135619146-a74b4c7b-fdea-432d-b91a-b1b04f110794.png)
![mania](https://user-images.githubusercontent.com/1329837/135619161-6fe24b91-0056-40e0-96f4-8f4a38028b4b.png)
On very dense maps (https://osu.ppy.sh/beatmapsets/39804#osu/129891 / https://osu.ppy.sh/beatmapsets/92190#mania/249346 respectively).

I've also made some "questionable" changes like making `GetPlayableBeatmap` virtual, instead of making `DifficultyCalculator` take in an `IBeatmap`/`IWorkingBeatmap` as the latter would be a breaking change. I'm not against such a breaking change but it should probably be part of a separate PR.  
This same reasoning goes for why I'm allocating a `PerformanceCalculator` per-judgement. Ideally these objects can be reused as they do everything inside the `Calculate()` method anyway, but doing that would again be a breaking change.

smoogipoo committed:
Initial PP counter implementation

smoogipoo committed:
Update styling

smoogipoo committed:
Fix possible nullrefs

smoogipoo committed:
Add counter to HUD

smoogipoo committed:
Merge branch 'master' into realtime-pp-display

smoogipoo committed:
Fix mania difficulty calculator crashing

smoogipoo committed:
Add xmldocs

smoogipoo committed:
Make Score internal

smoogipoo committed:
Split classes

smoogipoo committed:
Extra guard against no attributes

smoogipoo committed:
Increase test lenience

peppy commented on osu.Game.Tests/Visual/Gameplay/TestSceneBeatmapSkinFallbacks.cs:
```diff
@@ -84,18 +84,18 @@ protected bool AssertComponentsFromExpectedSource(SkinnableTarget target, ISkin
-                && info.Children.SequenceEqual(other.Children, new FuncEqualityComparer<SkinnableInfo>(almostEqual));
         }
 
+        private static bool almostEqual(SkinnableInfo info, SkinnableInfo other) =>
```
Is this failure somehow related to the changes in this PR? o.O

smoogipoo replied:
I haven't investigated too deeply, so I'm open for help here, but I believe it may be related to offsetting position by height of score by default?

bdach replied:
@smoogipoo's hunch is partially right, it is related to the offsetting. The rest of it is that the offsetting logic [is using `ScreenSpaceDrawQuad.Size.Y`](https://github.com/ppy/osu/blob/4e3d9da22d4422f2eb3b3e1f0c513ac95af45d4a/osu.Game/Skinning/DefaultSkin.cs#L87) rather than just plain `Height`. I'm not actually super sure why that would be required - I'd like to think all HUD skinnables should be in a common coordinate space.

peppy commented:
What's with the really large lag spike when adding a new counter in the skin editor in your video?

Also a bit unfortunate rewind doesn't work completely correctly, but I guess that's something we can work towards.

smoogipoo commented:
Unwind can work, just means binding to `OnJudgementReverted`. Didn't do that because I thought it was fine as is.

> What's with the really large lag spike when adding a new counter in the skin editor in your video?

It's doing a diffcalc.

peppy commented:
Ouch, re-calculating? At very least I'd want that to be async, but also means that's going to be induced during player load?

smoogipoo commented:
It's included in the player load, which is already async, yes. Not sure if there should be a separate level of async, but I did consider it.

peppy commented:
I get that, but it's increasing the length of async load. Didn't we have a caching component for difficulty calculation?

bdach commented:
Total/aggregate diffcalc yes, incremental diffcalc after every hitobject like this pull does no. You'd have to cache tha too and also compute it at song select no less, to avoid (or more like move) the load overhead.

smoogipoo commented:
Yeah I don't think we'd be caching it, but we could use the async process of the `BeatmapDifficultyCache` as an extra background load step.

peppy commented on osu.Game/Screens/Play/HUD/DefaultPerformancePointsCounter.cs:
```diff
@@ -0,0 +1,159 @@
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public class DefaultPerformancePointsCounter : RollingCounter<int>, ISkinnableDrawable
```
I'd probably drop the `Default` here as there's no legacy.

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -93,9 +93,9 @@ public abstract class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisab
+        public Ruleset GameplayRuleset { get; private set; }
 
-        protected GameplayBeatmap GameplayBeatmap { get; private set; }
+        public GameplayBeatmap GameplayBeatmap { get; private set; }
```
Is it feasible that we just expose `GameplayBeatmap` via DI and get the ruleset via that, rather than direction through `Player`? Really don't like having central classes like `Player` as a requirement for getting to these.

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -127,7 +127,7 @@ public abstract class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisab
         [Cached]
         [Cached(Type = typeof(IBindable<IReadOnlyList<Mod>>))]
-        protected new readonly Bindable<IReadOnlyList<Mod>> Mods = new Bindable<IReadOnlyList<Mod>>(Array.Empty<Mod>());
+        public new readonly Bindable<IReadOnlyList<Mod>> Mods = new Bindable<IReadOnlyList<Mod>>(Array.Empty<Mod>());
```
Does this need to become `public` if it's already exposed via DI?

peppy commented on osu.Game/Screens/Play/HUD/DefaultPerformancePointsCounter.cs:
```diff
@@ -0,0 +1,159 @@
+            }
+        }
+
+        private class GameplayWorkingBeatmap : WorkingBeatmap
```
maybe this is okay, but needs a TODO mentioning that it should be replaced and not a thing going forward.

peppy reviewed (commented):
Just an initial pass with thoughts

bdach commented on osu.Game/Screens/Play/HUD/DefaultPerformancePointsCounter.cs:
```diff
@@ -0,0 +1,159 @@
+            attribIndex = Math.Clamp(attribIndex, 0, timedAttributes.Length - 1);
+
+            var ppProcessor = gameplayRuleset.CreatePerformanceCalculator(timedAttributes[attribIndex].Attributes, player.Score.ScoreInfo);
+            Current.Value = (int)(ppProcessor?.Calculate() ?? 0);
```
`PerformanceStatistic` rounds rather than truncating:

https://github.com/ppy/osu/blob/1be96b42929ce410ba2e8b82a6ab55fad064f4c5/osu.Game/Screens/Ranking/Expanded/Statistics/PerformanceStatistic.cs#L46-L47

Not sure which is correct but they should at least be consistent.

smoogipoo committed:
Merge branch 'master' into realtime-pp-display

smoogipoo committed:
Make Mods readonly

smoogipoo committed:
Use GameplayState

smoogipoo committed:
Drop "default" prefix

smoogipoo committed:
Add todo

smoogipoo committed:
Share rounding with PerformanceStatistic

smoogipoo committed:
Fix missing CanBeNull

smoogipoo committed:
Pass score into GameplayState

smoogipoo committed:
Fix redundant default value

bdach commented:
Tests still need looking at - `TestScenePause.TestResumeWithResumeOverlaySkipped` is apparently failing windows-only (can't seem to repro on linux), while linux and macOS are still failing on the `BeatmapSkinFallbacks` one [mentioned above](https://github.com/ppy/osu/pull/14919#discussion_r720692450).

smoogipoo committed:
Fix skin fallbacks test

smoogipoo committed:
Further asyncify load process

smoogipoo committed:
Fix incorrect CancellationToken usage

Apparently I wrote the BDL system and don't know how this works. I
believe you need `CancellationToken?` or CanBeNull=true, however that doesn't actually
play well when actually using the token in code...

peppy commented on osu.Game/Screens/Play/GameplayState.cs:
```diff
@@ -30,7 +31,12 @@ public class GameplayState
+        /// <summary>
+        /// The gameplay score.
+        /// </summary>
+        public Score? Score { get; set; }
```
I'd really hope this is `readonly`, are you okay with me changing the flow in player to create the score early enough to change that? Seems like it should be possible.

smoogipoo replied:
Make sure to test it. Only reason I didn’t is because I didn’t want to look into how ReplayPlayer / Autoplay / Solo handle things…

peppy replied:
Yep. On a quick look they don't seem like they'd be affected but will test.

peppy committed:
Make `GameplayState.Score` immutable

peppy committed:
Revert `Player.Score` to `protected`

peppy commented on osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs:
```diff
@@ -178,10 +215,52 @@ static IEnumerable<Mod> createDifficultyAdjustmentModCombinations(ReadOnlyMemory
+
+            public double TotalBreakTime => baseBeatmap.TotalBreakTime;
+
+            public readonly List<HitObject> HitObjects = new List<HitObject>();
```
Just to confirm, the only special implementation detail of this class is that `HitObjects` is mutable?

smoogipoo replied:
Correct.

peppy commented on osu.Game/Screens/Play/HUD/PerformancePointsCounter.cs:
```diff
@@ -0,0 +1,173 @@
+            }
+        }
+
+        // Todo: This class shouldn't exist, but requires breaking changes to allow DifficultyCalculator to receive an IBeatmap.
```
I think at very least an issue should be created tracking this (or just pen in the follow up change sometime in the near future). Don't feel comfortable with it remaining in "todo never" inline todo land, especially since I'm actively trying to tighten the beatmap data flow..

peppy commented on osu.Game/Screens/Play/HUD/PerformancePointsCounter.cs:
```diff
@@ -0,0 +1,173 @@
+            var attribIndex = Array.BinarySearch(timedAttributes, 0, timedAttributes.Length, new TimedDifficultyAttributes(judgement.HitObject.GetEndTime(), null));
+            if (attribIndex < 0)
+                attribIndex = ~attribIndex - 1;
+            attribIndex = Math.Clamp(attribIndex, 0, timedAttributes.Length - 1);
```
is it actually possible for a value to be outside this range spec?

smoogipoo replied:
Yes. Specifically the max value.

> If this index is one greater than the upper bound of the array, there are no elements larger than value in the array

Meaning, `~attribIndex - 1` brings the value to `timedAttributes.Length`.

peppy committed:
Apply some code quality refactoring

peppy committed:
Adjust default location slightly, fix alignment of "pp" subtext

peppy commented:
Quite a pity this component has very ingrained dependencies meaning there is no test scene. Tempted to put in the effort to make one more..

peppy committed:
Fade display out during rewind (as the value displayed is no longer valid)

peppy commented:
@smoogipoo have pushed an initial batch of changes (mostly minor) for your review, but also looking to add a test atm.

smoogipoo committed:
Remove unnecessary dependency

smoogipoo commented:
Changes look good.

peppy committed:
Tidy up attribute retrieval code

peppy commented on osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs:
```diff
@@ -32,14 +42,41 @@ protected DifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
+            return CreateDifficultyAttributes(Beatmap, playableMods, skills, clockRate);
+        }
+
+        public IEnumerable<TimedDifficultyAttributes> CalculateTimed(params Mod[] mods)
```
Is there a reason to expose this as a public live `IEnumerable`? Seems quite unsafe, are you okay with returning an array here instead?

smoogipoo replied:
Sounds fine.

peppy replied:
I've changed this to a `List<>` as discussed. Please double-check.

peppy committed:
Add basic test scene

peppy committed:
Add easing to rolling counter value

peppy committed:
Avoid returning a live `IEnumerable`

peppy committed:
Add assert tests

peppy committed:
Fix display not displaying correctly after initial load

peppy committed:
Also dim counter during initial calculation phase

peppy committed:
Handle judgement reverts with actual display updates

peppy commented:
Ended up fully supporting reverting results since it seems to feel better.

smoogipoo commented on osu.Game.Tests/Visual/Gameplay/TestScenePerformancePointsCounter.cs:
```diff
@@ -0,0 +1,108 @@
+                scoreProcessor.RevertResult(new JudgementResult(new HitObject(), new OsuJudgement()));
+            });
+
+            AddUntilStep("counter decreased", () => counter.Current.Value < previousValue);
```
I'm ready for this test to start failing once the PP calculator changes... We'll see.

smoogipoo commented:
Still looks fine.
